feat: Integrate provenance raster generation into flood polygon pipeline

zackarno opened the pull request:
## Auto generate summary
Integrates automatic provenance raster generation into the `scripts/04_generate_flood_polygons.py` pipeline. The provenance raster tracks the last observation date for each flood pixel, enabling temporal analysis and visualization of flood evolution.

## Key Changes

### 1. Provenance Raster Generation ([scripts/04_generate_flood_polygons.py](scripts/04_generate_flood_polygons.py))
- Automatically generates provenance raster after polygon creation
- Uploads to `ds-flood-gfm/processed/provenance_raster/` in blob storage
- Uses same cache key naming convention as polygon outputs
- Format: `{cache_key}_provenance.tif`

### 2. Admin-Level Aggregation Helper ([src/ds_flood_gfm/datasources/gfm.py](src/ds_flood_gfm/datasources/gfm.py))
- Added `add_modal_provenance_to_admin()` function
- Calculates most common (modal) provenance date per admin boundary
- Uses `exactextract` for efficient zonal statistics
- Returns admin GeoDataFrame with `prov_date` and `prov_idx` columns

### 3. Parser Improvements
- Added `--n-search` argument to control STAC search window (default: 15 days)
- Better control over data availability vs processing time trade-off

### 4. Cache Key Naming Update
- Shortened filename format for large date ranges
- Format: `ISO3_YYYYMMDD_to_YYYYMMDD_Ndates_pop_mode`
- Example: `CUB_20251027_to_20251111_16dates_nopop_cumulative`

## Technical Details

### Lazy Evaluation
The provenance raster stays lazy until explicitly computed, minimizing memory usage:
```python
provenance_idx, date_mapping = create_provenance_raster(stack_flood_max, unique_dates)
prov_computed = provenance_idx.compute()  # Only computed when needed
```

### Upload Mechanism
- Uses `stratus.upload_cog_to_blob()` with xarray DataArray directly
- No temporary file creation needed
- Uploads as Cloud Optimized GeoTIFF (COG)

### Date Mapping
Integer indices map to observation dates:
```python
{0: '2025-11-05', 1: '2025-11-08', 2: '2025-11-10', 3: '2025-11-11', -1: 'No Data'}
```

## Testing

Tested successfully on:
- ✅ **Jamaica (JAM)** - Small AOI, 4 dates, cumulative mode
  - Output: `JAM_20251105_20251108_20251110_20251111_nopop_latest_provenance.tif`- **Currently testing on PHL**
## Example Usage

```bash
uv run python scripts/04_generate_flood_polygons.py \
  --iso3 JAM \
  --end-date 2025-11-12 \
  --n-latest 4 \
  --n-search 15 \
  --flood-mode cumulative
```

**Output:**
- Polygons: `ds-flood-gfm/processed/polygon/{cache_key}.shp.zip`
- Provenance: `ds-flood-gfm/processed/provenance_raster/{cache_key}_provenance.tif`

## Visualization

The provenance raster enables visualizations like:
- **Pixel-level:** Color each flood pixel by last observation date (red=oldest, green=newest)
- **Admin-level:** Show modal provenance date for each admin boundary
- **Temporal evolution:** Track how flood extent changes over time

See `demo_provenance_visualization.py` for examples (* will remove or move once pipeline is more thoroughly tested)

## Related Files

- `demo_provenance_visualization.py` - Demo script showing provenance visualization
- `flood_exposure.py` - Updated marimo notebook (added PHL support, minor formatting)

zackarno committed:
feat(parser): ✨ add `--n-search` argument to control search window

* Introduced `--n-search` parameter in the argument parser to specify the search window in days.
* Updated `query_gfm_stac` function call to utilize the new `n_search` argument.
* Enhanced logging to include the search window information.

zackarno committed:
feat: integrate optimized provenance raster generation into polygon pipeline

Add automatic provenance raster generation and upload to blob storage
in the scripts/04_generate_flood_polygons.py pipeline.

Changes:
- Generate provenance raster tracking last observation date per pixel
- Upload provenance raster as COG to blob storage using same cache key naming
- Add add_modal_provenance_to_admin() helper for admin-level aggregation
- Store in ds-flood-gfm/processed/provenance_raster/ with consistent naming

Technical details:
- Uses create_flood_composite with return_stack=True to get 3D stack
- Provenance stays lazy until .compute() for efficiency
- Uploads using stratus.upload_cog_to_blob with xarray DataArray directly
- Filename format: {cache_key}_provenance.tif

zackarno committed:
Add comprehensive dask diagnostics to identify computation bottlenecks

Added detailed logging and dask task graph analysis to diagnose why
large countries (PHL) fail during polygon generation.

Changes:
- Remove premature .sum() computation in create_flood_composite() to maintain
  lazy evaluation throughout the pipeline
- Add lazy evaluation tracking after each operation (stack, clip, groupby)
- Add dask task graph diagnostics before computation in raster_to_polygons():
  * Task count in computation graph
  * Estimated memory requirements
  * Chunk configuration
  * Actual computation time and memory used
- Add error handling and timing logs for polygon conversion

Key Findings from PHL Test:
- Successfully identifies the root cause: 34.8M dask tasks from accumulated
  lazy operations (not the polygon vectorization itself)
- Shows 40GB estimated memory for PHL (90,451 x 55,405 pixels)
- 285 STAC items over 16 dates creates massive computation graph

Remaining Issues:
- Large countries (PHL) still fail due to dask graph explosion
- Need optimization strategy for countries with >10M tasks:
  * Option 1: Persist intermediate results (compute stack_flood_max early)
  * Option 2: Rechunk data to reduce task count
  * Option 3: Process in spatial tiles
  * Option 4: Use distributed dask cluster for large countries

Works for: JAM ✅, HTI ✅, CUB ✅
Fails for: PHL ❌ (too many tasks)

zackarno committed:
feat(marimo): ✨ add runtime configuration for maximum output size

* Introduced a new configuration option `output_max_bytes` set to 15,000,000.
* This change allows for better control over the output size during processing.

zackarno committed:
feat: Add spatial tiling and chunking optimizations for large countries

Solves scaling problem for large countries like Philippines (PHL) through:

1. Chunking optimizations:
   - Increase chunksize to 2048 (4x larger)
   - Set rescale=False to skip unnecessary operations
   - Strategic rechunking: time=-1, spatial=4096x4096
   - Result: 34.8M → 8.8M tasks (75% reduction)

2. Spatial tiling for large countries:
   - Auto-detect large countries (PHL, IDN, BRA, USA, etc.)
   - Process in 2.0° tiles (configurable)
   - Per-tile: ~60k tasks, 0.12GB vs full country 8.8M tasks, 5GB
   - Result: 130x task reduction, 40x memory reduction per tile

3. Testing:
   - PHL: Successfully processes 45 tiles
   - JAM: Baseline test passes (475 polygons)

Known limitations:
- Provenance rasters not yet supported for tiled processing
- Note: Provenance may have pre-existing issues to investigate separately

zackarno committed:
fix(provenance): 🐛 cast provenance raster to `int16` for proper GeoTIFF encoding

* Ensure `prov_computed` is cast to `np.int16` to maintain correct data type.
* Updated logging to include the data type of the computed provenance raster.

hannahker commented on src/ds_flood_gfm/datasources/gfm.py:
```diff
@@ -98,22 +143,44 @@ def create_flood_composite(items: list, bbox: list, n_latest: int, mode: str = "
+
+    # RECHUNK: Consolidate time dimension and use larger spatial chunks
+    # This dramatically reduces task count for temporal operations
+    logger.info("  Rechunking for optimal temporal operations...")
```
Woah this is great to see. Chunking is one of those things that makes sense to me in theory but I've never really implemented well in practice. Nice!

hannahker commented on src/ds_flood_gfm/datasources/gfm.py:
```diff
@@ -182,9 +385,147 @@ def export_polygons(gdf: gpd.GeoDataFrame, output_path: Path, local=False, blob=
         output_path.parent.mkdir(parents=True, exist_ok=True)
         file_path = output_path.with_suffix('.shp')
-        gdf.to_file(file_path, driver='ESRI Shapefile') 
+        gdf.to_file(file_path, driver='ESRI Shapefile')
```
Maybe next PR we can switch this to geoparquet :)

zackarno replied:
yeah that would be great!

hannahker commented on src/ds_flood_gfm/datasources/gfm.py:
```diff
@@ -182,9 +385,147 @@ def export_polygons(gdf: gpd.GeoDataFrame, output_path: Path, local=False, blob=
+    logger.info(f"Adding modal provenance to admin level {admin_level} boundaries...")
+
+    # Write provenance raster to temporary file for exactextract
+    with tempfile.NamedTemporaryFile(suffix=".tif", delete=False) as tmp:
```
Not sure if this is more performant, but you should also be able to pass the raster in memory (eg. as an xr.DataArray object) to `exactextract`.

zackarno replied:
yeah  i think so, this is what you did in the Marimo app right?

i think  the current method is more performant, but  not sure if meaningfully so. Anyways, since this is working, im going to pull out this comment as an issue (#7) as a reminder for later

hannahker commented on scripts/04_generate_flood_polygons.py:
```diff
@@ -1,81 +1,187 @@
+        logger.info("GENERATING PROVENANCE RASTER")
+        logger.info("=" * 60)
+
+        provenance_idx, date_mapping = create_provenance_raster(
```
This date mapping gets a little bit lost in the logs. Would it be possible to store this in the COG metadata?

zackarno replied:
yes is pretty critically lost. It's reconstructable from the filenames and the band names (integers) however. I think an issues might be that stratus.upload_cog_to_blob()` may not support metadata, so we might want to modify that. Let's see what @copilot thinks just for kicks?

I was thinking a work around might just be to have a dedicated function `map_date(da)` that you can always run when you load the raster w/ xarray that would add the attribute.

zackarno replied:
doing some playing arround where im able to store some metadata as data-array attributes which then get preserved when written to a cog via stratus. Heres the output where you can see where the metadata lives and what it looks like in qgis:

<img width="1870" height="304" alt="image" src="https://github.com/user-attachments/assets/d22212ce-aae9-4a32-ae1e-9c7b0bde5b56" />

zackarno replied:
There might be better ways to handle metadata here, but [d8a7527](https://github.com/OCHA-DAP/ds-flood-gfm/pull/3/commits/d8a752772a8ac032114fb4f3de229d6abc3ed077) which produces the above in qgis and the below in python seems to be an improvement at least <img width="625" height="477" alt="image" src="https://github.com/user-attachments/assets/0973167f-44f3-46a9-b99c-41f0f4686ef3" />

hannahker commented on src/ds_flood_gfm/datasources/gfm.py:
```diff
@@ -182,9 +385,147 @@ def export_polygons(gdf: gpd.GeoDataFrame, output_path: Path, local=False, blob=
+    return provenance_idx, date_mapping
+
+
+def add_modal_provenance_to_admin(gdf_admin: gpd.GeoDataFrame,
```
Is this implemented somewhere? Would be nice to have this in the output shapefile!

zackarno replied:
yeah good call. It was previously implemented more from scratch (no imports) in `scripts/02_*` to make the admin level choropleth maps with admin-level provenance polygons overlaid. I just refactored the above function to make it more general as `add_modal_provenance_to_polygon()` and now implement it as part of the pipeline on flood polygons in `scripts/04_*` [ba22b3f](https://github.com/OCHA-DAP/ds-flood-gfm/pull/3/commits/ba22b3f295825fd9056e513bdfb4e47de0befaf6)

hannahker commented on OPTIMIZATION_STRATEGY.md:
SUPER interesting write up here. I wonder if we want to make use of the repo Wiki for docs like this?

zackarno replied:
do you mean a wiki somehow within this repo or is there a more general team wiki you are referring to? I feel like teh doc is now a bit specific (in fact i just made it more specific [b405502](https://github.com/OCHA-DAP/ds-flood-gfm/pull/3/commits/b405502753e30f357c3280d023001ea5e5f0ef7f)), but it would be good to distill some more general lessons from it

hannahker replied:
<img width="1125" height="124" alt="Screenshot 2025-11-27 at 11 45 47" src="https://github.com/user-attachments/assets/7f35a1c3-96d3-45a0-82b0-33ffb9c3201d" />

I just mean in the project Wiki on Github here! Not something we've used a lot of, but perhaps could be nice to more

hannahker reviewed (approved):
Really great!! The tiling looks well implemented and I'm so impressed with the optimization strategy. I tested with both PHL and JAM, and everything seems to work. 

Just some minor comments throughout. Possibly for a future PR, but we might want to add a bit to the marimo notebook to integrate the results of the provenance raster. As is, it seems like there's still some manual analysis that would need to be done to integrate.

zackarno committed:
feat(provenance): ✨ add metadata attributes to provenance raster

* Added `date_mapping` and `created_date` attributes to the provenance raster as GeoTIFF tags.
* Enhanced logging to confirm metadata addition and upload status.

zackarno committed:
feat(provenance): ✨ add provenance dates to flood polygons and refactor for reusability

Implements Hannah's PR review comments #4 and #5:
- Add provenance metadata (date_mapping, created_date) to COG files as xarray attrs
- Add provenance columns (prov_date, prov_idx) to flood polygon outputs
- Refactor add_modal_provenance_to_admin() → add_provenance_to_polygons() for generic use
- Eliminate duplicate exactextract code in script 02 by using refactored function

Changes:
- scripts/04_generate_flood_polygons.py: Integrate polygon provenance before export
- src/ds_flood_gfm/datasources/gfm.py: Generalize provenance function with optional id_col
- scripts/02_generate_affected_population_choropleths.py: Use shared function (2 occurrences)

Performance: Adds ~0.3s for 392 polygons, ~11s for provenance raster computation

zackarno committed:
docs: ✨ update OPTIMIZATION_STRATEGY.md with provenance integration and production metrics

Updates documentation to reflect current implementation:

New sections:
- Provenance Raster Performance: Details performance impact (~11.3s for Jamaica)
- Production Metrics: Real benchmarks from Jamaica and Philippines runs
- Updated Known Limitations: Clarifies provenance + tiling constraint and Phase 3 status

Key additions:
- Provenance integration details (int16 raster, xarray attrs, exactextract)
- Auto-detection of large countries (PHL, IDN, BRA, USA, CAN, RUS, CHN, AUS)
- Jamaica timing breakdown: 37s total with provenance
- Provenance distribution example (391 polygons from 2025-11-22)
- Clarified Phase 3 persist strategy not used (tiling more effective)

Addresses Hannah's PR comment about keeping optimization docs current.
